#24 Copy Offload - List Outstanding Offload Operations Request

bdevcich-hpe opened the pull request:
# Description

A new API request/response pair was added to support listing the outstanding data
movement requests for the copy-offload API:
    - DataMovementListRequest
    - DataMovementListResponse

This allows a user a way to query data movement requests for a given
Namespace and Workflow. These are required parameters. If there are no
data movement requests for the given namespace and workflow, an empty
list will be supplied in the response.

See `daemons/compute/Readme.md` for more detail regarding the API.

# Testing

Deployed `nnf-dm` daemon on dp1a. Used `client.go` to test creating a
number of requests, listing them, and removing them. Used global lustre
workflow as the destination along with a number of gfs2 workflows as the
source.

Test Cases:
- Verified client is able to retrieve requests with the matching
  workflow/namespace.
- Verified that suppling incorrect namespace/workflows result in an
  empty list for `uids`.
- Verified that only the requests with matching workflow/namespaces are
  listed when there are other reqeusts with another workflow.

Signed-off-by: Blake Devcich <blake.devcich@hpe.com>

NateRoiger commented on daemons/compute/server/servers/server_sim.go:
```diff
@@ -95,3 +95,17 @@ func (s *simulatedServer) Delete(ctx context.Context, req *pb.DataMovementDelete
+func (s *simulatedServer) List(ctx context.Context, req *pb.DataMovementListRequest) (*pb.DataMovementListResponse, error) {
+	if len(s.requests) > 0 {
+		// Get the uuids from the requests map
+		uids := make([]string, len(s.requests))
```
Since you allocate the array up front can you avoid the `append` call with just an assignment `uids[i]=key.String()`?

bdevcich-hpe replied:
Is that preferable to create another variable (`i`) and increment it over using append? Should I not be allocating the array up front?

NateRoiger replied:
ohh yeah, I forgot `range` on a map doesn't provide an index. OK - ignore my suggestion.

NateRoiger replied:
One more thing though - if you are appending to an array of `len(s.requests)`, then the first `len(s.requests)` are marked empty in the array; so you're really returning an array 2x the expected value. I think since these are empty strings that they get filtered out by protobuf, but it should still be corrected.

I had to prove this out myself; here's an example https://go.dev/play/p/VDUcXQICPBo

I think what you could do is allocate an array of _lenght_ 0 but _capacity_ = len(s.requests). So it would be `make([]string, 0, len(s.requests))`

NateRoiger commented on daemons/compute/server/servers/server_default.go:
```diff
@@ -308,6 +308,30 @@ Retry:
+	}
+
+	list := dmv1alpha1.RsyncNodeDataMovementList{}
+	if err := s.client.List(ctx, &list, opts...); err == nil {
```
Typically we code the exception case inside the if; so if you could invert this it would align better with other code
```go
if err := s.client.List(ctx, &list, opts...); err != nil {
  return nil, err
}
// good case
```

NateRoiger commented on daemons/compute/server/servers/server_default.go:
```diff
@@ -308,6 +308,30 @@ Retry:
+
+	list := dmv1alpha1.RsyncNodeDataMovementList{}
+	if err := s.client.List(ctx, &list, opts...); err == nil {
+		rsyncdm_uids := []string{}
```
Consider allocating the array upfront with `make([]string, len(list.Items))` and void the append (see comment in `server_sim.go`). Small optimization.

NateRoiger commented:
We've been providing an example C client - I think LLNL is using the compiled .a file directly so we should update that as well.

Form the _client-c folder, run `make protoc` then add an example List call to `client.go` (and maybe the equivalent of Free, i.e. FreeList()) - it could get ugly and I'm no CGo expert so you might need to do some research yourself.

You might need to `brew install protobuf-c` or something similar.

Reach out if you get stuck.

roehrich-hpe commented on daemons/compute/server/servers/server_default.go:
```diff
@@ -308,6 +308,30 @@ Retry:
+	// Only get the DMs that match the workflow and workflow namespace
+	opts := []client.ListOption{
+		client.MatchingLabels{"dws.cray.hpe.com/owner.namespace": req.Namespace},
+		client.MatchingLabels{"dws.cray.hpe.com/owner.name": req.Workflow},
```
```opts := []client.ListOption{
 		client.MatchingLabels{dwsv1alpha1.OwnerNameSpaceLabel: req.Namespace},
 		client.MatchingLabels{dwsv1alpha1.OwnerNameLabel: req.Workflow},
```

bdevcich-hpe replied:
Thanks for this!

NateRoiger commented on daemons/compute/server/servers/server_default.go:
```diff
@@ -317,8 +317,10 @@ func (s *defaultServer) List(ctx context.Context, req *pb.DataMovementListReques
-		rsyncdm_uids := []string{}
+	if err := s.client.List(ctx, &list, opts...); err != nil {
+		return nil, err
+	} else {
```
No need for this to be in an `else{}` block since the above is a return. This might seem like nit-picking, but it makes things easier to read - you just find your column and follow it down vertically,  knowing everything indented further is exception cases.

bdevcich-hpe replied:
Fixed.

bdevcich-hpe commented on daemons/compute/client-go/main.go:
```diff
@@ -67,46 +73,61 @@ func main() {
-		log.Printf("Data movement request deleted: %s %s", createResponse.GetUid(), deleteResponse.String())
+
+		// Print out the list again to verify deletes
+		getRequestsList(ctx, c, *workflow, *namespace, true)
```
I'll clean this up and remove the print

bdevcich-hpe commented on daemons/compute/client-go/main.go:
```diff
@@ -139,6 +160,35 @@ func getStatus(ctx context.Context, client pb.DataMoverClient, uid string, maxWa
+		Namespace: namespace,
+	})
+
+	if err != nil {
```
I'll do this for the other existing functions as well.

bdevcich-hpe replied:
Makes sense. I'll remove the function since we can just do it in `getRequestsList()`.

NateRoiger commented on daemons/compute/client-go/main.go:
```diff
@@ -139,6 +160,23 @@ func getStatus(ctx context.Context, client pb.DataMoverClient, uid string, maxWa
+		Namespace: namespace,
+	})
+
+	if err == nil {
```
Again, consider the if case the exception. Also, no need for a local variable
```go
if err != nil {
   log.Fatalf("...")
}

return rsp.GetUids()
```

bdevcich-hpe replied:
this should now look exactly like the other request/response functions.

NateRoiger commented on daemons/compute/client-go/main.go:
```diff
@@ -139,6 +160,23 @@ func getStatus(ctx context.Context, client pb.DataMoverClient, uid string, maxWa
 	return rsp, nil
 }
 
+func getRequestsList(ctx context.Context, client pb.DataMoverClient, workflow string, namespace string, print bool) []string {
```
This function is a little unorthodox - create/status/delete all return the raw protobuf response and an error; and in `func main()` is where the error check / `log.Fatalf` is. Consider following the existing pattern _or_ updating the other routines.

NateRoiger replied:
The `print bool` variable is unused and can be removed

bdevcich-hpe replied:
I'll make this consistent with the other functions.

I was originally attempting to reduce duplication by adding in the print statements since I wanted to print out the list before and after deletion and using two functions: `getRequestsList` and `listRequests`. The former to wrap in the error handling/printing and the former to return the raw response.

roehrich-hpe commented on daemons/compute/client-go/main.go:
```diff
@@ -67,45 +73,60 @@ func main() {
-	log.Printf("Data movement completed: %s", createResponse.GetUid())
+	// Get the list of Data Movement Requests
+	request_uids := getRequestsList(ctx, c, *workflow, *namespace, true)
+	log.Printf("List of Data movement requests: %v", strings.Join(request_uids, ", "))
```
If you're using %v,  then just use bare request_uids as the arg, and it'll print out some readable representation of the list.  It's not a bad way to go.

If you're using strings.Join(), then you're  getting a string back and you can print that with %s.

roehrich-hpe commented on .gitignore:
```diff
@@ -21,6 +21,8 @@ test.out
 daemons/compute/client-c/client
 daemons/compute/client-c/client.a
+daemons/compute/client-go/client-go
+daemons/compute/server/nnf-dm
```
These lines for client-go/client-go and server/nnf-dm are already in master.

bdevcich-hpe replied:
Thanks for pointing that out. I'll rebase and remove my additions.

roehrich-hpe reviewed (commented):
While you're here, add our copyright to daemons/compute/api/datamovement.proto.  You'll find it in hack/boilerplate.go.txt.  You might have to switch it to // for the comment block.

roehrich-hpe commented on daemons/compute/server/servers/server_default.go:
```diff
@@ -308,6 +308,30 @@ Retry:
 	}, nil
 }
 
+func (s *defaultServer) List(ctx context.Context, req *pb.DataMovementListRequest) (*pb.DataMovementListResponse, error) {
```
Hmm, where's the list for the NnfDataMovements that are used for the lustre-to-lustre copies?   I think you can let that be in a separate PR, if you prefer.

I would add it to the list that is returned from here.  The RsyncNodeDataMovement and NnfDataMovement resources are named differently, so they won't collide.

You can use the same `opts` list that you've already created here.  That'll work for the NnfDataMovement search as well.

NateRoiger replied:
I was ignoring this since v2 would deprecate rsync; I can add it then.

bdevcich-hpe replied:
Just to confirm - I need to get the lists for both DM types and then combine the uids in the result, correct?

NateRoiger replied:
Merge what you've done so far; we can come back to it in another PR.

bdevcich-hpe replied:
👍

roehrich-hpe commented:
During client-initiated L2L copy, the datamovement_controller creates a  PV/PVC pair.  While testing with this code, that pair  are not binding to each other.  I'm still trying to track this down, so  give me some time to dig into that.

roehrich-hpe commented:
> 

Well,  I'm not seeing the PV/PVC problem.  I must have had some mixed parts.

bdevcich-hpe commented:
Rebased (squashed) and retested. Here's the output of the client on compute-node-3:

```
$ ./client-go -namespace default -workflow gfs2 -source /mnt/nnf/6544/job/gfs2/test-blake -destination /lus/global -count 3
2022/08/04 15:28:15 Connecting to unix:///var/run/nnf-dm.sock
2022/08/04 15:28:15 Creating request 1 of 3...
2022/08/04 15:28:15 Data movement request created: default-gfs2-t9qws
2022/08/04 15:28:15 Data movement status status:SUCCESS
2022/08/04 15:28:16 Data movement status state:COMPLETED status:SUCCESS
2022/08/04 15:28:16 Data movement completed: default-gfs2-t9qws
2022/08/04 15:28:16 Creating request 2 of 3...
2022/08/04 15:28:16 Data movement request created: default-gfs2-jz7qm
2022/08/04 15:28:16 Data movement status status:SUCCESS
2022/08/04 15:28:17 Data movement status state:COMPLETED status:SUCCESS
2022/08/04 15:28:17 Data movement completed: default-gfs2-jz7qm
2022/08/04 15:28:17 Creating request 3 of 3...
2022/08/04 15:28:17 Data movement request created: default-gfs2-fvxj4
2022/08/04 15:28:17 Data movement status status:SUCCESS
2022/08/04 15:28:18 Data movement status state:COMPLETED status:SUCCESS
2022/08/04 15:28:18 Data movement completed: default-gfs2-fvxj4
2022/08/04 15:28:18 List of Data movement requests: default-gfs2-t9qws, default-gfs2-jz7qm, default-gfs2-fvxj4
2022/08/04 15:28:18 Deleting request: default-gfs2-t9qws
2022/08/04 15:28:18 Data movement request deleted: default-gfs2-t9qws status:DELETED
2022/08/04 15:28:18 Deleting request: default-gfs2-jz7qm
2022/08/04 15:28:18 Data movement request deleted: default-gfs2-jz7qm status:DELETED
2022/08/04 15:28:18 Deleting request: default-gfs2-fvxj4
2022/08/04 15:28:18 Data movement request deleted: default-gfs2-fvxj4 status:DELETED
2022/08/04 15:28:18 List of Data movement requests:

```
